NVML_ERROR_NOT_SUPPORTED exception

### 🐛 Describe the bug

Sometimes it can occur that NVML does not support monitoring queries to specific devices. Currently this leads to failing the startup phase.

### Error logs

```
2022-07-04T12:33:15,023 [ERROR] Thread-20 org.pytorch.serve.metrics.MetricCollector - Traceback (most recent call last):
  File "ts/metrics/metric_collector.py", line 27, in <module>
    system_metrics.collect_all(sys.modules['ts.metrics.system_metrics'], arguments.gpu)
  File "/usr/local/lib/python3.6/dist-packages/ts/metrics/system_metrics.py", line 91, in collect_all
    value(num_of_gpu)
  File "/usr/local/lib/python3.6/dist-packages/ts/metrics/system_metrics.py", line 72, in gpu_utilization
    statuses = list_gpus.device_statuses()
  File "/usr/local/lib/python3.6/dist-packages/nvgpu/list_gpus.py", line 67, in device_statuses
    return [device_status(device_index) for device_index in range(device_count)]
  File "/usr/local/lib/python3.6/dist-packages/nvgpu/list_gpus.py", line 67, in <listcomp>
    return [device_status(device_index) for device_index in range(device_count)]
  File "/usr/local/lib/python3.6/dist-packages/nvgpu/list_gpus.py", line 26, in device_status
    temperature = nv.nvmlDeviceGetTemperature(handle, nv.NVML_TEMPERATURE_GPU)
  File "/usr/local/lib/python3.6/dist-packages/pynvml/nvml.py", line 1956, in nvmlDeviceGetTemperature
    _nvmlCheckReturn(ret)
  File "/usr/local/lib/python3.6/dist-packages/pynvml/nvml.py", line 765, in _nvmlCheckReturn
    raise NVMLError(ret)
pynvml.nvml.NVMLError_NotSupported: Not Supported
```

### Installation instructions

pytorch/torchserve:latest-gpu


### Model Packaing

N/A

### config.properties

_No response_

### Versions

```
------------------------------------------------------------------------------------------
Environment headers
------------------------------------------------------------------------------------------
Torchserve branch: 

torchserve==0.6.0
torch-model-archiver==0.6.0

Python version: 3.6 (64-bit runtime)
Python executable: /usr/bin/python3

Versions of relevant python libraries:
future==0.18.2
numpy==1.19.5
nvgpu==0.9.0
psutil==5.9.1
requests==2.27.1
torch-model-archiver==0.6.0
torch-workflow-archiver==0.2.4
torchserve==0.6.0
wheel==0.30.0
**Warning: torch not present ..
**Warning: torchtext not present ..
**Warning: torchvision not present ..
**Warning: torchaudio not present ..

Java Version:


OS: N/A
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: N/A
CMake version: N/A
```

### Repro instructions

run:
```
torchserve --start --foreground --model-store model-store/ 
```

### Possible Solution

Deal with those exceptions.

--- a/ts/metrics/system_metrics.py
+++ b/ts/metrics/system_metrics.py
@@ -12,50 +12,51 @@ from ts.metrics.dimension import Dimension
 from ts.metrics.metric import Metric
 
 system_metrics = []
-dimension = [Dimension('Level', 'Host')]
+dimension = [Dimension("Level", "Host")]
 
 
 def cpu_utilization():
     data = psutil.cpu_percent()
-    system_metrics.append(Metric('CPUUtilization', data, 'percent', dimension))
+    system_metrics.append(Metric("CPUUtilization", data, "percent", dimension))
 
 
 def memory_used():
     data = psutil.virtual_memory().used / (1024 * 1024)  # in MB
-    system_metrics.append(Metric('MemoryUsed', data, 'MB', dimension))
+    system_metrics.append(Metric("MemoryUsed", data, "MB", dimension))
 
 
 def memory_available():
     data = psutil.virtual_memory().available / (1024 * 1024)  # in MB
-    system_metrics.append(Metric('MemoryAvailable', data, 'MB', dimension))
+    system_metrics.append(Metric("MemoryAvailable", data, "MB", dimension))
 
 
 def memory_utilization():
     data = psutil.virtual_memory().percent
-    system_metrics.append(Metric('MemoryUtilization', data, 'percent', dimension))
+    system_metrics.append(Metric("MemoryUtilization", data, "percent", dimension))
 
 
 def disk_used():
-    data = psutil.disk_usage('/').used / (1024 * 1024 * 1024)  # in GB
-    system_metrics.append(Metric('DiskUsage', data, 'GB', dimension))
+    data = psutil.disk_usage("/").used / (1024 * 1024 * 1024)  # in GB
+    system_metrics.append(Metric("DiskUsage", data, "GB", dimension))
 
 
 def disk_utilization():
-    data = psutil.disk_usage('/').percent
-    system_metrics.append(Metric('DiskUtilization', data, 'percent', dimension))
+    data = psutil.disk_usage("/").percent
+    system_metrics.append(Metric("DiskUtilization", data, "percent", dimension))
 
 
 def disk_available():
-    data = psutil.disk_usage('/').free / (1024 * 1024 * 1024)  # in GB
-    system_metrics.append(Metric('DiskAvailable', data, 'GB', dimension))
+    data = psutil.disk_usage("/").free / (1024 * 1024 * 1024)  # in GB
+    system_metrics.append(Metric("DiskAvailable", data, "GB", dimension))
+
 
 def gpu_utilization(num_of_gpu):
     """
-        Collect gpu metrics.
+    Collect gpu metrics.
 
-        :param num_of_gpu:
-        :return:
-        """
+    :param num_of_gpu:
+    :return:
+    """
     if num_of_gpu <= 0:
         return
 
@@ -69,19 +70,34 @@ def gpu_utilization(num_of_gpu):
 
     info = nvgpu.gpu_info()
     for value in info:
-        dimension_gpu = [Dimension('Level', 'Host'), Dimension("device_id", value['index'])]
-        system_metrics.append(Metric('GPUMemoryUtilization', value['mem_used_percent'], 'percent', dimension_gpu))
-        system_metrics.append(Metric('GPUMemoryUsed', value['mem_used'], 'MB', dimension_gpu))
+        dimension_gpu = [
+            Dimension("Level", "Host"),
+            Dimension("device_id", value["index"]),
+        ]
+        system_metrics.append(
+            Metric(
+                "GPUMemoryUtilization",
+                value["mem_used_percent"],
+                "percent",
+                dimension_gpu,
+            )
+        )
+        system_metrics.append(
+            Metric("GPUMemoryUsed", value["mem_used"], "MB", dimension_gpu)
+        )
 
     try:
         statuses = list_gpus.device_statuses()
     except pynvml.nvml.NVMLError_NotSupported:
-        logging.warning('gpu device monitoring not supported')
+        logging.warning("gpu device monitoring not supported")
         statuses = []
 
     for idx, status in enumerate(statuses):
-        dimension_gpu = [Dimension('Level', 'Host'), Dimension("device_id", idx)]
-        system_metrics.append(Metric('GPUUtilization', status['utilization'], 'percent', dimension_gpu))
+        dimension_gpu = [Dimension("Level", "Host"), Dimension("device_id", idx)]
+        system_metrics.append(
+            Metric("GPUUtilization", status["utilization"], "percent", dimension_gpu)
+        )
+
 
 def collect_all(mod, num_of_gpu):
     """
@@ -95,7 +111,10 @@ def collect_all(mod, num_of_gpu):
     members = dir(mod)
     for i in members:
         value = getattr(mod, i)
-        if isinstance(value, types.FunctionType) and value.__name__ not in ('collect_all', 'log_msg'):
+        if isinstance(value, types.FunctionType) and value.__name__ not in (
+            "collect_all",
+            "log_msg",
+        ):
             if value.__name__ == "gpu_utilization":
                 value(num_of_gpu)
             else:

--- a/ts/metrics/system_metrics.py
+++ b/ts/metrics/system_metrics.py
@@ -4,7 +4,10 @@ Module to collect system metrics for front-end
 import logging
 import types
 from builtins import str
+
 import psutil
+import pynvml
+
 from ts.metrics.dimension import Dimension
 from ts.metrics.metric import Metric
 
@@ -60,6 +63,7 @@ def gpu_utilization(num_of_gpu):
     # pylint: disable=import-outside-toplevel
     import nvgpu
     from nvgpu import list_gpus
+
     # pylint: enable=wrong-import-position
     # pylint: enable=import-outside-toplevel
 
@@ -69,7 +73,12 @@ def gpu_utilization(num_of_gpu):
         system_metrics.append(Metric('GPUMemoryUtilization', value['mem_used_percent'], 'percent', dimension_gpu))
         system_metrics.append(Metric('GPUMemoryUsed', value['mem_used'], 'MB', dimension_gpu))
 
-    statuses = list_gpus.device_statuses()
+    try:
+        statuses = list_gpus.device_statuses()
+    except pynvml.nvml.NVMLError_NotSupported:
+        logging.warning('gpu device monitoring not supported')
+        statuses = []
+
     for idx, status in enumerate(statuses):
         dimension_gpu = [Dimension('Level', 'Host'), Dimension("device_id", idx)]
         system_metrics.append(Metric('GPUUtilization', status['utilization'], 'percent', dimension_gpu))

